Assets/Mirror/Runtime/QuaternionReadWrite.cs

```diff
@@ -0,0 +1,348 @@
+                throw new System.ArgumentException("Quaternion must be normalized, Use 'Quaternion.identity' instead of 'default'");
+            }
+
+            int largestIndex = findLargestIndex(value);
+            Vector3 small = getSmallerDimensions(largestIndex, value);
+
+
+            if (precision == RotationPrecision.Full)
+            {
+                WriteQuaternionFull(writer, largestIndex, small);
+            }
+            else if (precision == RotationPrecision.Half)
+            {
+                WriteQuaternionHalf(writer, largestIndex, small);
+            }
+        }
+
+        /// <summary>
+        /// Writes Quaternion using QuaternionReadWrite.DefaultPrecision
+        /// </summary>
+        public static Quaternion ReadQuaternion(this NetworkReader reader)
+        {
+            return ReadQuaternion(reader, DefaultPrecision);
+        }
+        public static Quaternion ReadQuaternion(this NetworkReader reader, RotationPrecision precision)
+        {
+            if (precision == RotationPrecision.Full)
+            {
+                return ReadQuaternionFull(reader);
+            }
+            else if (precision == RotationPrecision.Half)
+            {
+                return ReadQuaternionHalf(reader);
+            }
+            else
+            {
+                return Quaternion.identity;
+            }
+        }
+
+        static void WriteQuaternionFull(NetworkWriter writer, int largestIndex, Vector3 small)
+        {
+            // 22 bits
+            int bitLength = 22;
+            uint a = scaleToUInt(small.x, bitLength);
+            uint b = scaleToUInt(small.y, bitLength);
+            uint c = scaleToUInt(small.z, bitLength);
+
+            // pack largestIndex info a
+            a |= ((uint)largestIndex) << 22;
+
+            writer.WriteByte((byte)a);
+            writer.WriteByte((byte)(a >> 8));
+            writer.WriteByte((byte)(a >> 16));
+
+            writer.WriteByte((byte)b);
+            writer.WriteByte((byte)(b >> 8));
+            writer.WriteByte((byte)(b >> 16));
+
+            writer.WriteByte((byte)c);
+            writer.WriteByte((byte)(c >> 8));
+            writer.WriteByte((byte)(c >> 16));
+        }
+        static Quaternion ReadQuaternionFull(NetworkReader reader)
+        {
+            Quaternion result;
+            uint a1 = reader.ReadByte();
+            uint a2 = reader.ReadByte();
+            uint a3 = reader.ReadByte();
+
+            uint b1 = reader.ReadByte();
+            uint b2 = reader.ReadByte();
+            uint b3 = reader.ReadByte();
+
+            uint c1 = reader.ReadByte();
+            uint c2 = reader.ReadByte();
+            uint c3 = reader.ReadByte();
+
+            uint a = a1 | (a2 << 8) | (a3 << 16);
+            uint b = b1 | (b2 << 8) | (b3 << 16);
+            uint c = c1 | (c2 << 8) | (c3 << 16);
+
+            uint largestIndex = a >> 22;
+            // removes largestIndex from a 
+            a &= ~(3u << 22);
```

Comment:
is the license fine for this code (not just this part, but all of it)?
or are there huge chunks of code copied from != public domain/MIT?

Reply:
I've implemented the code from the notes at the top. I used the GDC talk and this as references 
https://gist.github.com/gafferongames/bb7e593ba1b05da35ab6 which has `This source code is placed in the public domain.` at the top

Reply:
awesome, thanks!